[run-webkit-tests] Extra thread spawned for each worker

JonWBedard opened the pull request:
#### db09bfbcf7d0202ce90ff92340f51f4309501770
<pre>
[run-webkit-tests] Extra thread spawned for each worker
<a href="https://bugs.webkit.org/show_bug.cgi?id=247324">https://bugs.webkit.org/show_bug.cgi?id=247324</a>
rdar://101813483

Reviewed by Elliott Williams.

* Tools/Scripts/libraries/webkitcorepy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Bump version, import
NullContext before TaskPool.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/multiprocessing_utils.py: Added.
(Queue.__init__):
(Queue.close):
(Queue.__getstate__):
(Queue.__setstate__):
(Queue.get):
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/task_pool.py:
(_BiDirectionalQueue.__init__): Use multiprocessing_utils.Queue for Python 3, multiprocessing.Queue for Python 2.
(_BiDirectionalQueue.receive): Use Python 2/3 compatible Empty exception.
(_BiDirectionalQueue.close): Only invoke thread join if the Queue has the function.
(TaskPool.do): Use Python 2/3 compatible Empty exception.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/multiprocessing_utils_unittest.py: Added.
(QueueUnittest.test_basic):
(QueueUnittest.test_no_block):
(QueueUnittest.test_timeout):

Canonical link: <a href="https://commits.webkit.org/256262@main">https://commits.webkit.org/256262@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/234aad19b971f508b23a481c4047ebdfaf3509ad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95253 "4 style errors") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/4535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/28224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104847 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/165100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99243 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4519 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/33246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100919 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/81793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/87570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/98842 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/28224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38978 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/28224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/94230 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/28224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40735 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/81793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2085 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42719 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/28224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/multiprocessing_utils.py:
```diff
@@ -0,0 +1,121 @@
+# A combination of the multiprocessing module's SimpleQueue from Python 2.7.18 and 3.9.6, but with
+# a get() method that support the block and timeout arguments.
+class Queue(object):
+    Empty = BaseQueue.Empty
```
It looks like this is a modified copy of cpython's SimpleQueue (which helps answer some questions I had about implementation details): https://github.com/python/cpython/blob/main/Lib/multiprocessing/queues.py#L337

Why copy SimpleQueue instead of subclassing it, or by creating a wrapper object that manages a SimpleQueue behind a lock? Also, IANAL but I'm not sure that we can establish copyright over this.

JonWBedard replied:
Python 2 version is MIT licensed (and their license is very nearly word-for-word identical to ours), forgot to include the original author, making that change.

And I didn't sub-class `SimpleQueue` because it's kind of painful to subclass because `multiprocessing` exports `SimpleQueue` as a function. Also, different versions of Python have slightly different `SimpleQueue` implementations enough that it's easier to just do this ourselves than try and understand what is different about each Python version.  The `close()` function, for example, is new in Python 3.9.

emw-apple replied:
> Also, different versions of Python have slightly different `SimpleQueue` implementations enough that it's easier to just do this ourselves than try and understand what is different about each Python version. The `close()` function, for example, is new in Python 3.9.

Sure, but it means our implementation may easily break in future python versions.

> And I didn't sub-class `SimpleQueue` because it's kind of painful to subclass because `multiprocessing` exports `SimpleQueue` as a function.

Maybe there are details I'm missing, but I think we could use `multiprocessing.queues.SimpleQueue` directly when subclassing.

JonWBedard replied:
I just spent about an hour working on this.

It works for Python 3, but not for Python 2 (because the `SimpleQueue` definition in Python 2 is kind of nonsense). I think we are going to have to spin this ourselves if we want something that is vaguely sane.

aproskuryakov replied:
Do we need Python 2 support here? Perhaps it's good enough to retain the existing code for Python 2, or just switch to 3 entirely.

JonWBedard replied:
We could fix this bug exclusively for Python 3. The reason I didn't is because `webkitcorepy` definitely needs to support Python 2 in general, even if `run-webkit-tests` does not. It also doesn't do a ton for code complexity, it will only save us maybe 2 dozen lines (and move around the places we have to consider Python 2, we will need to, for example, bring back and conditionalize the `join_thread` calls in `TaskPool`

emw-apple commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/multiprocessing_utils.py:
```diff
@@ -0,0 +1,77 @@
+
+    def close(self):
+        self._closed = True
+        self._reader.close()
+        self._writer.close()
```
nit: can we do `super(Queue, self).close()` here?

emw-apple commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/multiprocessing_utils.py:
```diff
@@ -0,0 +1,77 @@
+
+    def __init__(self):
+        self._closed = False
+        super(Queue, self).__init__(ctx=context._default_context)
```
nit: I believe `multiprocessing.get_context()` is the documented way to get the default context.

emw-apple commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/multiprocessing_utils.py:
```diff
@@ -0,0 +1,77 @@
+                raise self.Empty
+            return _ForkingPickler.loads(self._reader.recv_bytes())
+        finally:
+            self._rlock.release()
```
I think we could initialize `self._rlock` with a `multiprocessing.RLock`, which would simply our `get` implementation because we could make all paths call into super's `get()`. But no need to bikeshed on it further unless you want to, thanks for making this change.

JonWBedard committed:
[run-webkit-tests] Extra thread spawned for each worker
https://bugs.webkit.org/show_bug.cgi?id=247324
rdar://101813483

Reviewed by Elliott Williams.

* Tools/Scripts/libraries/webkitcorepy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Bump version, import
NullContext before TaskPool.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/multiprocessing_utils.py: Added.
(Queue.__init__):
(Queue.close):
(Queue.__getstate__):
(Queue.__setstate__):
(Queue.get):
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/task_pool.py:
(_BiDirectionalQueue.__init__): Use multiprocessing_utils.Queue for Python 3, multiprocessing.Queue for Python 2.
(_BiDirectionalQueue.receive): Use Python 2/3 compatible Empty exception.
(_BiDirectionalQueue.close): Only invoke thread join if the Queue has the function.
(TaskPool.do): Use Python 2/3 compatible Empty exception.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/multiprocessing_utils_unittest.py: Added.
(QueueUnittest.test_basic):
(QueueUnittest.test_no_block):
(QueueUnittest.test_timeout):

Canonical link: https://commits.webkit.org/256262@main
